src/client/auth/oidc.rs

```diff
@@ -23,44 +24,117 @@ use crate::{
 #[derive(Clone)]
-pub struct Callbacks {
-    inner: Arc<CallbacksInner>,
+enum Callback {
+    Machine(Arc<CallbackInner>),
+    Human(Arc<CallbackInner>),
 }
 
-impl Callbacks {
-    /// Create a new instance with a token request callback.
-    pub fn new<F>(on_request: F) -> Self
+// TODO RUST-1497: These methods will be public
+#[allow(dead_code)]
+impl Callback {
+    /// Create a new instance with a human token request callback.
+    pub fn human<F>(callback: F) -> State
     where
-        F: Fn(IdpServerInfo, RequestParameters) -> BoxFuture<'static, Result<IdpServerResponse>>
+        F: Fn(CallbackContext) -> BoxFuture<'static, Result<IdpServerResponse>>
             + Send
             + Sync
             + 'static,
     {
-        Self {
-            inner: Arc::new(CallbacksInner {
-                on_request: Box::new(on_request),
-            }),
+        State {
+            callback: Callback::Human(Arc::new(CallbackInner {
+                f: Box::new(callback),
+            })),
+            cache: Arc::new(RwLock::new(Cache::default())),
+        }
+    }
+
+    /// Create a new instance with a machine token request callback.
+    pub fn machine<F>(callback: F) -> State
+    where
+        F: Fn(CallbackContext) -> BoxFuture<'static, Result<IdpServerResponse>>
+            + Send
+            + Sync
+            + 'static,
+    {
+        State {
+            callback: Callback::Machine(Arc::new(CallbackInner {
+                f: Box::new(callback),
+            })),
+            cache: Arc::new(RwLock::new(Cache::default())),
         }
     }
 }
 
-impl std::fmt::Debug for Callbacks {
+impl std::fmt::Debug for Callback {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Callbacks").finish()
+        f.debug_struct("Callback").finish()
     }
 }
 
-struct CallbacksInner {
-    on_request: Box<
-        dyn Fn(IdpServerInfo, RequestParameters) -> BoxFuture<'static, Result<IdpServerResponse>>
-            + Send
-            + Sync,
-    >,
-    // on_refresh: Option<Box<dyn Fn(&IdpServerInfo) -> IdpServerResponse + Send + Sync>>,
+pub struct CallbackInner {
+    f: Box<dyn Fn(CallbackContext) -> BoxFuture<'static, Result<IdpServerResponse>> + Send + Sync>,
+}
+
+#[derive(Debug)]
+pub struct Cache {
+    refresh_token: Option<String>,
+    access_token: Option<String>,
+    token_gen_id: i32,
+    last_call_time: Instant,
 }
 
-#[derive(Debug, Deserialize)]
+impl Clone for Cache {
+    fn clone(&self) -> Self {
+        Self {
+            refresh_token: self.refresh_token.clone(),
+            access_token: self.access_token.clone(),
+            token_gen_id: self.token_gen_id,
+            last_call_time: self.last_call_time,
+        }
+    }
+}
+
+impl Default for Cache {
+    fn default() -> Self {
+        Self {
+            refresh_token: None,
+            access_token: None,
+            token_gen_id: 0,
+            last_call_time: Instant::now(),
```

Comment:
It's a little unusual for subsequent `default()` calls to produce different values.  Does this need to have a `Default` impl or can this be a `new()` method?

Reply:
`new` would be fine, I'm pretty sure.